File: 2022/4b.py
import re
import sys

def overlap(left, right):
    return left[0] <= right[1] and left[1] >= right[0]

dash_or_comma = re.compile(r"-|,")

print(sum(
    1
    for e1s, e1e, e2s, e2e in (
        map(int, dash_or_comma.split(line))
        for line in
        sys.stdin.read().split()
    )
    if overlap((e1s, e1e), (e2s, e2e))
        or overlap((e2s, e2e), (e1s, e1e))
))

File: 2022/4b.hs
import Data.Char (isDigit)

overlap :: (Int, Int, Int, Int) -> Bool
overlap (e1s, e1e, e2s, e2e)
  | e1s <= e2e && e1e >= e2s = True
  | e2s <= e1e && e2e >= e1s = True
  | otherwise = False

extractInts :: String -> (Int, Int, Int, Int)
extractInts s = (read e1s, read e1e, read e2s, read e2e)
  where (e2e, _) = span isDigit rest1
        (e2s, _:rest1) = span isDigit rest2
        (e1e, _:rest2) = span isDigit rest3
        (e1s, _:rest3) = span isDigit s

processInput :: [String] -> String
processInput = show . length . filter overlap . map extractInts

main = interact ((++ "\n") . processInput . lines)
